### ISSUE TITLE
Alternative approach to Leaflet props and events

### ISSUE THREAD

**Issue by mikeu** [vue-leaflet/vue-leaflet#2]
@DonNicoJs last night I pushed some changes to [my fork's dev branch](https://github.com/mikeu/vue-leaflet/tree/c09efd250b9934bc1d12aeca748541e4cf34e97c) that use `context.attrs` (composition api's combination of `this.$attrs` and `this.$listeners`) to allow the majority of the props and events to be set on the underlying Leaflet objects without having to define props on the Vue components, wrap them into an object, and pass that through.

I haven't pushed it very far yet so there might be a good reason not to use this sort of approach that I haven't thought of or encountered yet, but you can see initially at least how little code and effort is required to get `LMap`, `LTileLayer`, and `LMarker` up and running. I'm interested in your thoughts on the approach.

# Outline

The basic idea is to get the non-prop attributes passed to the component in `setup(props, { attrs })`, and split them into two sets: one whose names start with `"on"` and whose values are functions, and another with all the rest. The names of the first set are then changed slightly by removing `"on"` from the beginning and converting the first letter to lowercase. For example, if
```js
attrs == {
  onZoomend: fn,
  maxZoom: 16,
  center: [0, 0]
}
```
then we end up with
```js
listenersSet = { zoomend: fn }; // Set 1
optionsSet = { maxZoom: 16, center: [0, 0] }; // Set 2
```
(If the component is used with `<l-map @zoomend="fn" ...>`, then the composition api automatically converts that to `onZoomend: fn` in the `attrs` object. So to use the same event names that Leaflet users expect, we just have to undo that change.)

Now it's just a matter of instantiating a new Leaflet object, using standard Vue `props` for any non-options constructor arguments, and passing the second set of options down. For example,
```js
layerMarker = L.marker(props.latLng, optionsSet);
```
and using Leaflet's `DomEvent` to bind its event map,
```js
DomEvent.on(layerMarker, listenersSet);
```

## Advantages
* Significantly less code, both to write and maintain.
* Handles any new Leaflet options and events seamlessly, should future versions include them, and doesn't add code that expects or requires any particular Leaflet version in either direction.
* ...?

## Disadvantages
* Doesn't provide the user with autocomplete options for prop or event names in their IDE (although Vue 3's [`emits`](https://v3.vuejs.org/api/options-data.html#emits) property would allow getting around that somewhat, though without the future-proofing advantage above).
* Doesn't allow validation functionality for most Leaflet options.
* ...?

**Comment by DonNicoJs** [comment]
@mikeu I thought about this approach as well! And to be honest `less code to write and maintain` is a very appealing thing, in this codebase, I am already doing something very similar to grab the listeners actually (since `$listeners` is gone ) 

What worries me is that we lose the props `type` safety and the ability to write custom validators if we think we need them.
We are also removing the pattern that every prop is `reactive` and everything that goes in `options` is not which I think may lead to more confusion.

What do you think ? What is your stance between pro and cons?

**Comment by mikeu** [comment]
@DonNicoJs  I'm certainly not particularly committed to this approach at the moment, I just started using it on a new branch and found some things flowed _very_ easily, so liked that aspect. I was on the fence about it though, which is why I figured it was worth getting some more input on!

I agree with you about `props` types and validators—losing those wouldn't feel great. I had vaguely hoped that there might be a way to leverage the `@types/leaflet` package to provide design-time type safety (or at least type hinting for IDEs), but I don't have enough experience with it to be sure that's possible. In some ways I think this is a question of whether we want vue-leaflet to feel closer to a Vue mapping library in its own right, with its own defined API that happens to use Leaflet under the hood (and so mirror it quite closely), or if we want it to be a more lightweight wrapper, exposing the Leaflet API directly through Vue components, without taking responsibility for how that API looks in or interacts with Vue. Both have their pluses and minuses, for sure.

As for reactivity though, it certainly still seems to be possible to push changes in the non-prop attrs down to Leaflet from the `onUpdated` lifecycle hook. We'd essentially need the same sort logic as in `propsBinder`, but instead of setting watchers it would take the place of those watchers and just call the appropriate function(s) on each change. Something like
```js
onUpdated(() => {
  const neAttrs = getNonEventAttrs(attrs);
  const options = {};
  for (const key in neAttrs) {
    const setMethodName = 'set' + capitalizeFirstLetter(key);
    if (customSetters[setMethodName]) {
      customSetters[setMethodName](neAttrs[key]);
    } else if (leafletRef.value[setMethodName]) {
      leafletRef.value[setMethodName](neAttrs[key]);
    } else {
      options[key] = neAttrs[key];
    }
  }
  setOptions(leafletRef.value, options);
});
```

---

At this point I mostly just wanted to float the idea. If you've already thought of it and decided it's not the right approach here, I'm happy to follow your lead. There are a lot of benefits for consumers of the library to having as many props explicitly defined as possible.

I might play around with it a bit longer, flesh out the `onUpdated` idea above for example, but the more I think about it, the more I worry about losing the props types and validators too...

**Comment by DonNicoJs** [comment]
@mikeu Thanks for the in-depth answer!  Be sure to take a look at how I rewrote propsBinder in this repo! We could totally use it as is for `context.attrs` 

I am mostly worried that we are going to lose the props types and also the auto-generated documentation, I've seen sooo much confusion already and the same questions over and over.

The idea of using TS has already floated around in another issue, but for the moment I am not feeling confident that I can pull of the full rewrite in TS.

What do you think if we continue with the props approach for now? I would like to enlist your help here so that we can clear as many components as possible.

During this weekend I want to finalize Map, TileLayer, and Marker, and publish a pre-alpha package

Then I will open issue for all the remaining components and we can tackle them in parallel, WDYT?

**Comment by mikeu** [comment]
@DonNicoJs that sounds like a plan, props it is!

I won't make any substantive changes while you're working on that. Will wait to hear from you for the next steps, and in the meantime just keep experimenting with the composition API a bit more.

> Be sure to take a look at how I rewrote propsBinder in this repo! We could totally use it as is for context.attrs

Based on my limited exposure to Vue 3 so far, I'm not actually sure we can—in what I tried this morning, a watch added to `context.attrs.x` never fired. I had to use the `onUpdated` lifecycle hook to do something when one changed (and of course it fires when _anything_ changes, and doesn't tell you what changed). I very well might have just missed the documentation on what I should have done differently though, and that's really a different issue anyway :)
